owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -357,6 +376,17 @@ public void reportBlockCompletion(FileBlockInfo blockInfo) {
         uncompletedPieces.put(pieceIndex, status);
       } else {
         completedPieces.add(pieceIndex);
+        // Advertise with Have message for more bandwidth
+        for (Peer leecher : leechers) {
+          if (peers.containsKey(leecher) && !leecher.isAmChoked() && leecher.isPeerInterested()) {
+            PeerConnectionContext conn = peers.get(leecher);
+            try {
+              conn.peerConnector.writeMessage(new HaveMessage(pieceIndex));
+            } catch (IOException e) {
+              log.error(e);
+            }
```

Comment:
Advertising to all peers for every piece is extremely inefficient... Also, you probably want to advertise periodically instead of on piece completion.